src/main/java/org/folio/rest/impl/CheckinReceivePiecesHelper.java

```diff
@@ -186,43 +181,40 @@ private void checkIfAllPiecesFound(List<String> pieceIds, List<Piece> pieces) {
-   *
-   * @return {@link CompletableFuture} which holds map with PO line id as key
-   *         and list of corresponding pieces as value
-   */
-  CompletableFuture<Map<String, List<Piece>>> updateInventoryItems(Map<String, List<Piece>> piecesGroupedByPoLine) {
-    // Collect all piece records with non-empty item ids. The result is a map
-    // with item id as a key and piece record as a value
-    Map<String, Piece> piecesWithItems = collectPiecesWithItemId(piecesGroupedByPoLine);
+  private PoLine searchPoLineById(List<PoLine> poLines, String polineId) {
+    for (PoLine  poline: poLines){
+      if (poline.getId().equals(polineId)){
+        return poline;
+      }
+    } return null;
+  }
 
-    // If there are no pieces with ItemId, continue
-    if (piecesWithItems.isEmpty()) {
-      return completedFuture(piecesGroupedByPoLine);
+  private CompletableFuture<Boolean> updateHoldingAndItem(JsonObject item, Piece piece, PoLine poLine, String locationId) {
+    if (holdingUpdateOnCheckinReceiveRequired(piece, locationId, poLine)) {
+      return inventoryHelper.getOrCreateHoldingsRecord(poLine.getInstanceId(), locationId)
+        .thenCompose(holdingId -> {
+          item.put(ITEM_HOLDINGS_RECORD_ID, holdingId);
+          return receiveInventoryItemAndUpdatePiece(item, piece);
+        })
+        .exceptionally(t -> {
+          logger.error("Cannot create holding for specified piece location ", piece.getLocationId());
+          addError(piece.getPoLineId(), piece.getId(), ITEM_UPDATE_FAILED.toError());
+          return false;
+        });
+    } else {
+      return receiveInventoryItemAndUpdatePiece(item, piece);
     }
+  }
 
-    return getItemRecords(piecesWithItems)
-      .thenCompose(items -> {
-        List<CompletableFuture<Boolean>> futuresForItemUpdates = new ArrayList<>();
-        for (JsonObject item : items) {
-          String itemId = item.getString(ID);
-          Piece piece = piecesWithItems.get(itemId);
-          futuresForItemUpdates.add(receiveInventoryItemAndUpdatePiece(item, piece));
-        }
+  private boolean ifLocationNotProcessed(String locationId) {
+    return processedLocations.add(locationId);
+  }
 
-        return collectResultsOnSuccess(futuresForItemUpdates)
-          .thenApply(results -> {
-            if (logger.isDebugEnabled()) {
-              long successQty = results.stream()
-                .filter(result -> result)
-                .count();
-              logger.debug("{} out of {} inventory item(s) successfully updated", successQty, results.size());
-            }
-            return piecesGroupedByPoLine;
-          });
-      });
+  private boolean holdingUpdateOnCheckinReceiveRequired(Piece piece, String locationId, PoLine poLine) {
+    return isHoldingsUpdateRequired(poLine.getEresource(), poLine.getPhysical())
+      && StringUtils.isNotEmpty(locationId)
+      && !StringUtils.equals(locationId, piece.getLocationId())
+      && ifLocationNotProcessed(locationId);
```

Comment:
I believe this check is required in the `updateInventoryItems` method with additional update mentioned above.
Let's assume you are going to receive 3 pieces:
- `piece A` is linked to `instance A` and was originally linked to `location A`
- `piece B` is linked to `instance A` and was originally linked to `location A`
- `piece C` is linked to `instance B` and was originally linked to `location A`
The receiving request is
- `piece A` is now linked to `location B`
- `piece B` is now linked to `location B`
- `piece C` is now linked to `location B`
This means the holding records needs to be processed independently